test/TensorFlow/sese_loop_canonicalization.swift

```diff
@@ -0,0 +1,206 @@
+// CHECK:                       {condition Header: {{.*}}
+// CHECK:                         block {{.*}}
+// CHECK:                         block {{.*}}}
+// CHECK:                       block {{.*}}]>
+// CHECK:                   {condition Header: {{.*}}
+// CHECK:                     block {{.*}}
+// CHECK:                     block {{.*}}}]}
+// CHECK:               block {{.*}}]>
+// CHECK:           block {{.*}}]
+// CHECK:         block {{.*}}}
+// CHECK:       block {{.*}}]>
+// CHECK:   block {{.*}}]
+// CHECK: --- XLA CFG Canonicalize end
+
+// This example checks that the loop structure is preserved when we clone
+// the body of a loop during canonicalization.
+//expected-warning @+1 {{value implicitly copied to the host}}
+public func testLoopWithDoublyNestedLoopsRequiringCloning(
+	_ breakIndex:Int32, _ repetitions: Int32) -> Tensor<Int32> {
+	var result = Tensor<Int32>(0)
+	for _ in 1...repetitions {
+		var i = result
+		// On all exits of this loop, `result` is assigned a value that is computed
+		// in the loop. Therefore, we cannot avoid undefs unless we clone the body.
+		repeat {
+			// expected-warning @+1 {{value implicitly copied to the host}}
+			if i > breakIndex {
+				// + 1 so that this is not a simple assignment. 
+				// (Simple assignments won't trigger cloning.)
+				result = i + 1
+				break
+			}
+
+			// Crazy way to do k = i^2 so that we have a doubly loop when we unroll.
+			var k = Tensor<Int32>(0)
+			// expected-warning @+1 {{value implicitly copied to the host}}
+			while k < i {
+				var w = Tensor<Int32>(0)
+				// expected-warning @+1 {{value implicitly copied to the host}}
+				while w < i {
+					w += 1
+				}
+				k += 1
+			}
+			i += 1
+			// + k so that this is not a simple assignment. 
+			// (Simple assignments won't trigger cloning.)
+			result = i + k
+			// expected-warning @+1 {{value implicitly copied to the host}}
+		} while i <= 100
+		result += 3
+	}
+	// expected-note @+1 {{value used here}}
+	return result
+}
+
+
+// CHECK-LABEL: --- XLA CFG Loops Before Canonicalize: {{.*}}LoopWithDoublyNestedLoopsRequiringCloning{{.*}}
+// CHECK: Loop at depth 1 containing: {{.*}}
+// CHECK:     Loop at depth 2 containing: {{.*}}
+// CHECK:         Loop at depth 3 containing: {{.*}}
+// CHECK:             Loop at depth 4 containing: {{.*}}
+
+// CHECK-LABEL: --- XLA CFG Loops Before Canonicalize end
+// CHECK: Loop at depth 1 containing: {{.*}}
+// CHECK:     Loop at depth 2 containing: {{.*}}
+// CHECK:         Loop at depth 3 containing: {{.*}}
+// CHECK:             Loop at depth 4 containing: {{.*}}
+// CHECK:     Loop at depth 2 containing: {{.*}}
+// CHECK:         Loop at depth 3 containing: {{.*}}
+
+//-- Check the structure of the SESE region. (Note the cloned doubly nested loop.)
+// CHECK-LABEL: --- XLA CFG Canonicalize: {{.*}}LoopWithDoublyNestedLoopsRequiringCloning{{.*}}
+// CHECK: [sequence
+// CHECK:   <while Preheader: {{.*}}, Header: {{.*}}, exit: {{.*}}
```

Comment:
similar questions to above:
1. does this loop correspond to "loop at depth 2", not 1?
2. should line 180 be indented to match the indentation here?

Reply:
See the related comments above.